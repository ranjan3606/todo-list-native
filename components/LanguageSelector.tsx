import React from 'react';
import { View, Text, StyleSheet, TouchableOpacity } from 'react-native';
import { FontAwesome } from '@expo/vector-icons';
import { Colors } from '@/constants/Colors';
import { useTranslation } from '@/i18n';

interface LanguageSelectorProps {
  colorScheme: 'light' | 'dark';
}

export function LanguageSelector({ colorScheme }: LanguageSelectorProps) {
  const { language, setLanguage, availableLanguages, t } = useTranslation();

  return (
    <View style={styles.languageOptions}>
      {availableLanguages.map((lang) => (
        <TouchableOpacity 
          key={lang.code}
          testID={`language-option-${lang.code}`}
          style={[
            styles.languageOption, 
            language === lang.code && styles.selectedOption,
            { backgroundColor: Colors[colorScheme].cardBackground }
          ]}
          onPress={() => setLanguage(lang.code)}
        >
          <Text 
            style={[
              styles.optionText, 
              { color: Colors[colorScheme].text }
            ]}
          >
            {lang.name}
          </Text>
          {language === lang.code && (
            <FontAwesome 
              name="check" 
              size={18} 
              color={Colors.primary} 
              style={styles.checkIcon} 
            />
          )}
        </TouchableOpacity>
      ))}
    </View>
  );
}

const styles = StyleSheet.create({
  languageOptions: {
    flexDirection: 'column',
    marginTop: 10,
  },
  languageOption: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    padding: 15,
    borderRadius: 8,
    marginBottom: 10,
  },
  selectedOption: {
    borderWidth: 2,
    borderColor: Colors.primary,
  },
  optionText: {
    fontWeight: '500',
    fontSize: 16,
  },
  checkIcon: {
    marginLeft: 10,
  }
});
