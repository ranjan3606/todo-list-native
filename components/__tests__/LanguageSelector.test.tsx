import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { LanguageSelector } from '../LanguageSelector';

jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(() => Promise.resolve(null)),
  setItem: jest.fn(() => Promise.resolve(null)),
  removeItem: jest.fn(() => Promise.resolve(null)),
  clear: jest.fn(() => Promise.resolve(null)),
  getAllKeys: jest.fn(() => Promise.resolve([])),
  multiGet: jest.fn(() => Promise.resolve([])),
  multiSet: jest.fn(() => Promise.resolve(null)),
  multiRemove: jest.fn(() => Promise.resolve(null)),
}));

jest.mock('expo-haptics', () => ({
  impactAsync: jest.fn(),
  ImpactFeedbackStyle: { Light: 'light' }
}));

jest.mock('@expo/vector-icons', () => ({
  FontAwesome: function MockFontAwesome(props: any) {
    return null; // Simple mock that doesn't try to load any fonts
  }
}));

jest.mock('@/constants/Colors', () => ({
  Colors: {
    light: { cardBackground: '#fff', text: '#000' },
    dark: { cardBackground: '#222', text: '#fff' },
    primary: '#007aff',
  }
}));

jest.mock('@/i18n', () => ({
  useTranslation: () => ({
    language: 'en',
    setLanguage: jest.fn(),
    availableLanguages: [
      { code: 'en', name: 'English' },
      { code: 'es', name: 'Español' },
      { code: 'fr', name: 'Français' }
    ],
    t: (key: string) => key
  })
}));

describe('LanguageSelector', () => {
  it('renders correctly (snapshot)', () => {
    const { toJSON } = render(<LanguageSelector colorScheme="light" />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('displays all available languages', () => {
    const { getByText } = render(<LanguageSelector colorScheme="light" />);
    
    // All three languages should be present
    expect(getByText('English')).toBeTruthy();
    expect(getByText('Español')).toBeTruthy();
    expect(getByText('Français')).toBeTruthy();
  });

  it('marks current language as selected', () => {
    const { getByText, getByTestId } = render(<LanguageSelector colorScheme="light" />);
    
    // English should be selected (wrapper should have a border)
    const englishOption = getByTestId('language-option-en');
    expect(englishOption.props.style).toEqual(
      expect.objectContaining({ borderColor: expect.anything() })
    );
  });

  it('calls setLanguage when a language option is clicked', () => {
    const mockSetLanguage = jest.fn();
    
    // Override the mock implementation just for this test
    jest.spyOn(require('@/i18n'), 'useTranslation').mockImplementation(() => ({
      language: 'en',
      setLanguage: mockSetLanguage,
      availableLanguages: [
        { code: 'en', name: 'English' },
        { code: 'es', name: 'Español' },
        { code: 'fr', name: 'Français' }
      ],
      t: (key: string) => key
    }));
    
    const { getByText } = render(<LanguageSelector colorScheme="light" />);
    
    // Click on Spanish
    fireEvent.press(getByText('Español'));
    
    // setLanguage should be called with 'es'
    expect(mockSetLanguage).toHaveBeenCalledWith('es');
    
    // Click on French
    fireEvent.press(getByText('Français'));
    
    // setLanguage should be called with 'fr'
    expect(mockSetLanguage).toHaveBeenCalledWith('fr');
  });

  it('applies correct theme styles based on colorScheme prop', () => {
    // Test with light mode
    const { rerender, getByTestId } = render(<LanguageSelector colorScheme="light" />);
    
    const lightOption = getByTestId('language-option-en');
    expect(lightOption.props.style).toEqual(
      expect.objectContaining({ backgroundColor: '#fff' })
    );
    
    // Re-render with dark mode
    rerender(<LanguageSelector colorScheme="dark" />);
    
    const darkOption = getByTestId('language-option-en');
    expect(darkOption.props.style).toEqual(
      expect.objectContaining({ backgroundColor: '#222' })
    );
  });
});