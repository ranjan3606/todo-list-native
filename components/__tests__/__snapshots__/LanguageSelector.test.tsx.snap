// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSelector renders correctly (snapshot) 1`] = `
<View
  style={
    {
      "flexDirection": "column",
      "marginTop": 10,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderColor": "#007aff",
        "borderRadius": 8,
        "borderWidth": 2,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 10,
        "opacity": 1,
        "padding": 15,
      }
    }
    testID="language-option-en"
  >
    <Text
      style={
        [
          {
            "fontSize": 16,
            "fontWeight": "500",
          },
          {
            "color": "#000",
          },
        ]
      }
    >
      English
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 8,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 10,
        "opacity": 1,
        "padding": 15,
      }
    }
    testID="language-option-es"
  >
    <Text
      style={
        [
          {
            "fontSize": 16,
            "fontWeight": "500",
          },
          {
            "color": "#000",
          },
        ]
      }
    >
      Español
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 8,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 10,
        "opacity": 1,
        "padding": 15,
      }
    }
    testID="language-option-fr"
  >
    <Text
      style={
        [
          {
            "fontSize": 16,
            "fontWeight": "500",
          },
          {
            "color": "#000",
          },
        ]
      }
    >
      Français
    </Text>
  </View>
</View>
`;
